product_attribute_types/models/product.py

```diff
@@ -0,0 +1,58 @@
+# -*- encoding: utf-8 -*-
+##############################################################################
+#
+#    This program is free software: you can redistribute it and/or modify
+#    it under the terms of the GNU General Public License as published by
+#    the Free Software Foundation, either version 3 of the License, or
+#    (at your option) any later version.
+#
+#    This program is distributed in the hope that it will be useful,
+#    but WITHOUT ANY WARRANTY; without even the implied warranty of
+#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+#    GNU General Public License for more details.
+#
+#    You should have received a copy of the GNU General Public License
+#    along with this program.  If not, see http://www.gnu.org/licenses/.
+#
+##############################################################################
+
+from openerp import models, fields, _
+import openerp.addons.decimal_precision as dp
+
+
+class ProductAttribute(models.Model):
+    _inherit = "product.attribute"
+
+    attr_type = fields.Selection(required=True, selection=[
+        ('select', 'Select'),
+        ('range', 'Range'),
+        ('numeric', 'Numeric')], string="Type", default='select')
+
+
+class ProductAttributeLine(models.Model):
+    _inherit = "product.attribute.line"
+
+    required = fields.Boolean('Required')
+    default = fields.Many2one('product.attribute.value', 'Default')
+    attr_type = fields.Selection(string='Type', store=False,
+                                 related='attribute_id.attr_type')
+
+
+class ProductAttributeValue(models.Model):
+    _inherit = "product.attribute.value"
+
+    attr_type = fields.Selection(string='Type',
+                                 related='attribute_id.attr_type')
+    numeric_value = fields.Float('Numeric Value',
+                                 digits=dp.get_precision('Product Attribute'))
+    min_range = fields.Float('Min',
+                             digits=dp.get_precision('Product Attribute'))
+    max_range = fields.Float('Max',
+                             digits=dp.get_precision('Product Attribute'))
+
+    _sql_constraints = [
+        ('range_check',
+         'CHECK (min_range < max_range)',
```

Comment:
This constraint cannot be added on a database with demo data, and if it could be added, it would break adding values for attributes of other types than range because in that case min_range == max_range == 0.0.